hail/python/hail/matrixtable.py

```diff
@@ -811,15 +811,18 @@ def key_rows_by(self, *keys, **named_keys) -> 'MatrixTable':
         :class:`.MatrixTable`
         """
 
-        keys = get_select_exprs("MatrixTable.key_rows_by",
+        key_fields = get_select_exprs("MatrixTable.key_rows_by",
                              keys, named_keys, self._row_indices,
                              protect_keys=False)
 
-        row = self._rvrow.annotate(**keys)
+        new_row = self._rvrow.annotate(**key_fields)
+        base, cleanup = self._process_joins(new_row)
 
-        return self._select_rows("MatrixTable.key_rows_by",
-                                 row,
-                                 list(keys.keys()))
+        return cleanup(MatrixTable(
+            self._jvds
+                .keyRowsBy([])
+                .selectRows(str(new_row._ir))
```

Comment:
actually, there needs to be logic here which looks for key truncation, I think

Reply:
Other than changing `self` to `base`, this is identical to `Table.key_by` (which you reviewed!). If you recall, I had that logic in at first, but then simplified the python side by adding rules to the optimizer.

If we're truncating the key, then all new key fields will be references to existing key fields. So the `MatrixMapRows` will be a no-op, and will get optimized away, and then the two consecutive `MatrixKeyRowsBy` nodes will be collapsed into one. That one remaining node handles all the logic of comparing existing keys to new keys.